From e288fafeacd3a3c7443987f0f5763ab5a097a06a Mon Sep 17 00:00:00 2001
From: Razvan A Lupusoru <razvan.a.lupusoru@intel.com>
Date: Mon, 11 Feb 2013 08:03:06 -0800
Subject: Dalvik: Fix String.regionMatches error message

BZ: 86267

When regionMatches returns NullPointerException because the other string is
null, the message included needs to be "string == null".

Category: device-enablement
Domain: AOSP-Dalvik-Runtime
Origin: internal
Upstream-Candidate: yes

Change-Id: Idd9a12d01a1f04bc4dd5d7c5c3669ee5666c0c9a
Orig-MCG-Change-Id: I58045372e7b0c00d80a67d933755b781b84df575
Signed-off-by: Razvan A Lupusoru <razvan.a.lupusoru@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/native/java_lang_String.cpp |   12 +++++++++++-
 1 files changed, 11 insertions(+), 1 deletions(-)

diff --git a/vm/native/java_lang_String.cpp b/vm/native/java_lang_String.cpp
index ed690b4..7c05735 100644
--- a/vm/native/java_lang_String.cpp
+++ b/vm/native/java_lang_String.cpp
@@ -87,7 +87,17 @@ static void String_regionMatches(const u4* args, JValue* pResult)
 
     // Check if this and the other string are null
     if (thisString == NULL || otherString == NULL) {
-        dvmThrowNullPointerException(NULL);
+        // Technically thisString should never be null when we get here but
+        // we should handle it instead of crashing later in this function
+
+        const char * message = NULL;
+
+        // When the otherString is null, we need to include a message
+        if (otherString == NULL) {
+            message = "string == null";
+        }
+
+        dvmThrowNullPointerException(message);
         RETURN_VOID();
     }
 
-- 
1.7.4.1

